### PR TITLE
Task statedb evm imple 2

### PR DESCRIPTION
**Pull request by boqiu**
statedb implements EVM interface: refund, suicide and log

---
This change is [Reviewable](https://reviewable.io/reviews/seeleteam/go-seele/160)

### CODE REVIEW

File: core/state/statedb_evm.go [hunk 1]
```diff
@@ -64,13 +64,12 @@ func (s *Statedb) GetCodeSize(address common.Address) int {
 
 // AddRefund refunds the specified gas value
 func (s *Statedb) AddRefund(gas uint64) {
-	// @todo
+	s.refund += gas
```
Comment (WTer) [inline core/state/statedb_evm.go]: gas can be zero?

File: core/state/statedb_evm.go [hunk 2]
```diff
@@ -95,7 +94,7 @@ func (s *Statedb) Suicide(address common.Address) bool {
 	}
 
 	stateObj.SetAmount(new(big.Int))
-	// @todo mark the state object as suicided
+	stateObj.suicided = true
```
Comment (WTer) [inline core/state/statedb_evm.go]: do we need to add check to verify whether this account already suicided?

File: core/state/statedb_evm.go (hunk 1 above)
Comment (boqiu) [reply core/state/statedb_evm.go]: EVM use this API to refund gas, as for the gas value, generally > 0.

File: core/state/stateobject.go [hunk 3]
```diff
@@ -24,19 +24,22 @@ type Account struct {
...
+	address  common.Address
+	addrHash common.Hash
 
 	account      Account
 	dirtyAccount bool
 
 	code      []byte // contract code
 	dirtyCode bool
+
+	suicided bool
```
Comment (WTer) [inline core/state/stateobject.go]: nice to add some comments for this var

File: core/state/statedb_evm.go (hunk 2 above)
Comment (boqiu) [reply core/state/statedb_evm.go]: just mark a state object as suicided, and physically delete from trie/DB when commit the statedb DB.

File: core/state/stateobject.go (hunk 3 above)
Comment (boqiu) [reply core/state/stateobject.go]: Done

**Comment by WTer** [comment]
Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[core/state/statedb_evm.go, line 67 at r1](https://reviewable.io/reviews/seeleteam/go-seele/160#-LCCp7MJCMtovuXI2dXK:-LCCp7MJCMtovuXI2dXL:bou9xip) ([raw file](https://github.com/seeleteam/go-seele/blob/bb68f883ed14fdf3a4a86a001d5dab2f8414e235/core/state/statedb_evm.go#L67)):*
> ```
> s.refund += gas
> ```

---


*Comments from [Reviewable](https://reviewable.io/reviews/seeleteam/go-seele/160)*

**Comment by lucusfly** [comment]
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[core/state/stateobject.go, line 133 at r2](https://reviewable.io/reviews/seeleteam/go-seele/160#-LCDGJSM5eVA5Xq4CgTJ:-LCDGJSM5eVA5Xq4CgTK:b-jd3n5p) ([raw file](https://github.com/seeleteam/go-seele/blob/807f5bdfa0fd2d4abaee2e01c325a43565e0f644/core/state/stateobject.go#L133)):*
> ```Go
> }
> 
> func (s *StateObject) empty() bool {
> ```

what does empty mean? add comments?

---


*Comments from [Reviewable](https://reviewable.io/reviews/seeleteam/go-seele/160)*

**Comment by lucusfly** [comment]
Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/seeleteam/go-seele/160)*

**Comment by boqiu** [comment]
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[core/state/stateobject.go, line 133 at r2](https://reviewable.io/reviews/seeleteam/go-seele/160#-LCDGJSM5eVA5Xq4CgTJ:-LCDjHbDAyOv9ra-yal-:b-896fix) ([raw file](https://github.com/seeleteam/go-seele/blob/807f5bdfa0fd2d4abaee2e01c325a43565e0f644/core/state/stateobject.go#L133)):*
<details><summary><i>Previously, lucusfly (lucus) wrote…</i></summary><blockquote>

what does empty mean? add comments?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/seeleteam/go-seele/160)*

**Pull request merged**